### PR TITLE
Fix Brave Ads are not shown until user foregrounds app on mobile

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/30927

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: ios/browser/api/ads/brave_ads.mm [hunk 1]
```diff
@@ -475,13 +475,15 @@ - (void)applicationDidBecomeActive {
...
   adsClientNotifier->NotifyBrowserDidEnterForeground();
+  adsClientNotifier->NotifyBrowserDidBecomeActive();
 }
 
 - (void)applicationDidBackground {
   if (![self isAdsServiceRunning]) {
     return;
   }
   adsClientNotifier->NotifyBrowserDidEnterBackground();
+  adsClientNotifier->NotifyBrowserDidResignActive();
```
Comment (kylehickinson) [inline ios/browser/api/ads/brave_ads.mm:486]: Does the order of this matter? I.e. in the did become active you do `enter foreground` → `become active`.  On this side its `entered background` → `resign active` but that's technically backwards

Comment (tmancey) [reply ios/browser/api/ads/brave_ads.mm:485]: It should not matter, however I have made the order the same, and improved a couple other parts of this PR. Thanks

**Review by aseren — APPROVED** [review]

**Comment by btlechowski** [comment]
Verification passed on Brave v1.54.53 on Lenovo TB-8506FS (Android 11.0)

<details>

Verified `Browser did enter foreground`

```
21:32:07.984  V  [VERBOSE1:browser_manager.cc(74)] Browser did enter foreground
```

Verified `Start serving notification ads at regular intervals`

```
21:32:08.080  V  [VERBOSE1:notification_ad_serving.cc(56)] Start serving notification ads at regular intervals
```

Verified `Browser did become active`
```
21:32:08.109  V  [VERBOSE1:browser_manager.cc(54)] Browser did become active
```

Verified notification ad was served, viewed and confirmation was sent
```

21:33:08.131  V  [VERBOSE1:eligible_notification_ads_v2.cc(38)] Get eligible notification ads
21:33:08.149  V  [VERBOSE2:exclusion_rules_base.cc(120)] creativeSetId 45ea5952-37ca-41f0-88c2-4b546b0e6f4b has exceeded the perDay frequency cap
21:33:08.154  V  [VERBOSE1:eligible_notification_ads_v2.cc(124)] 2 eligible ads out of 12 ads
21:33:08.154  V  [VERBOSE1:notification_ad_handler.cc(168)] Opportunity arose to serve a notification ad
21:33:08.155  V  [VERBOSE1:notification_ad_serving.cc(125)] Found 1 eligible ads
21:33:08.155  V  [VERBOSE1:notification_ad_serving.cc(189)] Served notification ad
21:33:08.155  V  [VERBOSE1:notification_ad_handler.cc(185)] Served notification ad:
                   placementId: 309f28e5-4ca5-410b-bd82-1f8a713486f4
                   creativeInstanceId: 4b73d655-e818-4b9e-91ea-44aa42ad5aba
                   creativeSetId: a5e82762-1339-458c-9719-b2ad793f248b
                   campaignId: d1e25388-b362-4f30-bd4b-36b8205e59e5
                   advertiserId: a34de75d-5654-4331-8713-8f4d4e2b6bf7
                   segment: untargeted
                   title: Vivamus sagittis lacus vel aug
                   body: Duis mollis, est non commodo luctus, nisi erat porttitor lig
                   targetUrl: https://www.apple.com/
21:33:08.221  V  [VERBOSE1:notification_ad_serving.cc(161)] Maybe serve notification ad in 0 hours, 6 minutes, 0 seconds at 21:39:08.184
21:33:08.234  V  [VERBOSE3:notification_ad_handler.cc(196)] Served notification ad with placement id 309f28e5-4ca5-410b-bd82-1f8a713486f4 and creative instance id 4b73d655-e818-4b9e-91ea-44aa42ad5aba
21:33:08.249  V  [VERBOSE3:notification_ad_handler.cc(205)] Viewed notification ad with placement id 309f28e5-4ca5-410b-bd82-1f8a713486f4 and creative instance id 4b73d655-e818-4b9e-91ea-44aa42ad5aba
21:33:08.268  V  [VERBOSE3:account.cc(248)] Successfully processed deposit for ad_notification with creative instance id 4b73d655-e818-4b9e-91ea-44aa42ad5aba and view valued at 0.01
21:33:08.270  V  [VERBOSE1:confirmations.cc(108)] Confirming opted-in view for ad_notification with transaction id 54f0dee1-6c76-411d-a934-631eaa5d0239 and creative instance id 4b73d655-e818-4b9e-91ea-44aa42ad5aba
21:33:08.285  V  [VERBOSE1:redeem_opted_in_confirmation.cc(71)] Redeem opted-in confirmation
21:33:08.293  V  [VERBOSE1:redeem_opted_in_confirmation.cc(92)] CreateConfirmation
21:33:08.293  V  [VERBOSE2:redeem_opted_in_confirmation.cc(93)] POST /v3/confirmation/{transactionId}/{credential}
21:33:08.295  V  [VERBOSE6:redeem_opted_in_confirmation.cc(98)] URL Request:
                   URL: https://anonymous.ads.bravesoftware.com/v3/confirmation/54f0dee1-6c76-411d-a934-631eaa5d0239/eyJzaWduYXR1cmUiOiJyTWw1VHZwWi9vZU1HMjZFZ09DR0JTbkJ6QjA2blF0ZERSV0NBZ3BMQ2F3N2ZKVUdaQytBY3htSDcyU0JMVGhZV2FsT3ZHREMxSTZZaEhwTFZQbkQ0Zz09IiwidCI6Ikc5SGZkaHMrRi9PaTNYN0RvLzVpYkhOK044dGVxRGNCdE5CdVB2MTM2Nk9pZkFrcFhyVk1EQ0N3T2RNbS9YcEZERzBCSmRUMWQxbUxaMEJCZGhoWFpnPT0ifQ==
                   Content: {"blindedPaymentTokens":["uiZlt1Efmh4OoHY7XUQLCit0aHLGX+YZDpAL6/0v20c="],"buildChannel":"canary","catalog":[{"id":"c3fc80ed9df0ec3577dee088aa9f2ec42541b552"}],"createdAtTimestamp":"2023-06-12T19:00:00.000Z","creativeInstanceId":"4b73d655-e818-4b9e-91ea-44aa42ad5aba","platform":"android","publicKey":"hLaC3RQw5++hx8zaT+47CA8r8cOlAm+ygeky/ci6T1s=","rotating_hash":"/nPXQ3y5Bk5I8hiR0OBa3yIlI9gHaRYcsRb8NT5J7sU=","segment":"untargeted","studies":[{"group":"DefaultAdNotificationsPerHour=10/MaximumAdNotificationsPerDay=100/MaximumInlineContentAdsPerHour=6/MaximumInlineContentAdsPerDay=20/AdServingVersion=2","name":"BraveAds.AdServingStudy"}],"systemTimestamp":"2023-06-12T19:00:00.000Z","transactionId":"54f0dee1-6c76-411d-a934-631eaa5d0239","type":"view","versionNumber":"114.0.5735.110"}
                   Content Type: application/json
                   Method: kPost
21:33:09.393  V  [VERBOSE1:redeem_opted_in_confirmation.cc(112)] OnCreateConfirmation
21:33:09.394  V  [VERBOSE6:redeem_opted_in_confirmation.cc(114)] URL Response:
                   URL: https://anonymous.ads.bravesoftware.com/v3/confirmation/54f0dee1-6c76-411d-a934-631eaa5d0239/eyJzaWduYXR1cmUiOiJyTWw1VHZwWi9vZU1HMjZFZ09DR0JTbkJ6QjA2blF0ZERSV0NBZ3BMQ2F3N2ZKVUdaQytBY3htSDcyU0JMVGhZV2FsT3ZHREMxSTZZaEhwTFZQbkQ0Zz09IiwidCI6Ikc5SGZkaHMrRi9PaTNYN0RvLzVpYkhOK044dGVxRGNCdE5CdVB2MTM2Nk9pZkFrcFhyVk1EQ0N3T2RNbS9YcEZERzBCSmRUMWQxbUxaMEJCZGhoWFpnPT0ifQ==
                   Response Status Code: 201
                   Response: {"id":"54f0dee1-6c76-411d-a934-631eaa5d0239","payload":{},"createdAt":"2023-06-12T19:40:27.475Z","type":"view","modifiedAt":"2023-06-12T19:40:27.475Z","creativeInstanceId":"4b73d655-e818-4b9e-91ea-44aa42ad5aba"}
21:33:09.394  V  [VERBOSE1:redeem_opted_in_confirmation.cc(127)] FetchPaymentToken
21:33:09.396  V  [VERBOSE6:redeem_opted_in_confirmation.cc(131)] URL Request:
                   URL: https://anonymous.ads.bravesoftware.com/v3/confirmation/54f0dee1-6c76-411d-a934-631eaa5d0239/paymentToken
                   Method: kGet
21:33:09.678  V  [VERBOSE1:redeem_opted_in_confirmation.cc(145)] OnFetchPaymentToken
21:33:09.679  V  [VERBOSE6:redeem_opted_in_confirmation.cc(147)] URL Response:
                   URL: https://anonymous.ads.bravesoftware.com/v3/confirmation/54f0dee1-6c76-411d-a934-631eaa5d0239/paymentToken
                   Response Status Code: 202
                   Response: {"error":"Payment Tokens are not ready","statusCode":202}
				   
```				   
				   
Verified `Browser did resign active`
```
21:44:19.957  V  [VERBOSE1:browser_manager.cc(56)] Browser did resign active
```

Verified `Browser did enter background`
```
21:44:19.959  V  [VERBOSE1:browser_manager.cc(76)] Browser did enter background
```

Verified `Stop serving notification ads at regular intervals`				   
```
21:44:19.962  V  [VERBOSE1:notification_ad_serving.cc(68)] Stop serving notification ads at regular intervals	
```

Verified ads are not served when `Stop serving notification ads at regular intervals`		

</details>

**Comment by btlechowski** [comment]
Verification passed on

Brave | 1.54.53 Chromium: 114.0.5735.110 (Official Build) nightly (64-bit) 
-- | --
Revision | 1c828682b85bbc70230a48f5e345489ec447373e-refs/branch-heads/5735_90@{#13}
OS | Ubuntu 18.04 LTS

<details>

Verified `Browser did enter foreground`

```

[1979:1979:0612/220655.684391:VERBOSE1:browser_manager.cc(74)] Browser did enter foreground

```

Verified `Browser did become active`
```
[1979:1979:0612/220655.654489:VERBOSE1:browser_manager.cc(54)] Browser did become active
```

Verified notification ad was served, viewed and confirmation was sent
```
[1979:1979:0612/220816.525958:VERBOSE1:eligible_notification_ads_v2.cc(38)] Get eligible notification ads
[1979:1979:0612/220816.535292:VERBOSE1:eligible_notification_ads_v2.cc(124)] 1 eligible ads out of 1 ads
[1979:1979:0612/220816.535727:VERBOSE1:notification_ad_handler.cc(168)] Opportunity arose to serve a notification ad
[1979:1979:0612/220816.536321:VERBOSE1:notification_ad_serving.cc(125)] Found 1 eligible ads
[1979:1979:0612/220816.536705:VERBOSE1:notification_ad_serving.cc(189)] Served notification ad
[1979:1979:0612/220816.537045:VERBOSE1:notification_ad_handler.cc(185)] Served notification ad:
  placementId: 814ea858-ef49-4622-9ef3-ee167151e61f
  creativeInstanceId: 4cda3ae7-1f70-4bca-b361-ca01d6a8f1ed
  creativeSetId: 5f8c4787-8431-49a5-9d09-bba40f8ad5f4
  campaignId: d1e25388-b362-4f30-bd4b-36b8205e59e5
  advertiserId: a34de75d-5654-4331-8713-8f4d4e2b6bf7
  segment: untargeted
  title: HTML Conversion brave.com
  body: HTML Conversion brave.com
  targetUrl: https://www.apple.com/
[1979:1979:0612/220816.543313:VERBOSE6:ads_service_impl.cc(819)] Timeout notification ad with placement id 814ea858-ef49-4622-9ef3-ee167151e61f in 120 s
[1979:1979:0612/220816.576506:VERBOSE3:notification_ad_handler.cc(196)] Served notification ad with placement id 814ea858-ef49-4622-9ef3-ee167151e61f and creative instance id 4cda3ae7-1f70-4bca-b361-ca01d6a8f1ed
[1979:1979:0612/220816.576696:VERBOSE9:client_state_manager.cc(476)] Saving client state
[1979:1979:0612/220816.586735:VERBOSE9:client_state_manager.cc(485)] Successfully saved client state
[1979:1979:0612/220816.597066:VERBOSE3:notification_ad_handler.cc(205)] Viewed notification ad with placement id 814ea858-ef49-4622-9ef3-ee167151e61f and creative instance id 4cda3ae7-1f70-4bca-b361-ca01d6a8f1ed
[1979:1979:0612/220816.598992:VERBOSE9:client_state_manager.cc(476)] Saving client state
[1979:1979:0612/220816.604878:VERBOSE9:client_state_manager.cc(485)] Successfully saved client state
[1979:1979:0612/220816.637691:VERBOSE3:account.cc(248)] Successfully processed deposit for ad_notification with creative instance id 4cda3ae7-1f70-4bca-b361-ca01d6a8f1ed and view valued at 0.01
[1979:1979:0612/220816.649709:VERBOSE1:confirmations.cc(108)] Confirming opted-in view for ad_notification with transaction id 898875bd-81c3-435a-bddf-f9798cfcfcd4 and creative instance id 4cda3ae7-1f70-4bca-b361-ca01d6a8f1ed
[1979:1979:0612/220816.656054:VERBOSE9:confirmation_state_manager.cc(184)] Saving confirmations state
[1979:1979:0612/220816.658801:VERBOSE1:redeem_opted_in_confirmation.cc(71)] Redeem opted-in confirmation
[1979:1979:0612/220816.667254:VERBOSE1:redeem_opted_in_confirmation.cc(92)] CreateConfirmation
		   
```				   
				   
Verified `Browser did resign active`
```
[1979:1979:0612/220819.756317:VERBOSE1:browser_manager.cc(56)] Browser did resign active

```

Verified `Browser did enter background`
```
[1979:1979:0612/220819.760157:VERBOSE1:browser_manager.cc(76)] Browser did enter background

```

Verified ads are not served
```
[1979:1979:0612/221632.487471:VERBOSE1:user_idle_detection.cc(36)] User is active after 5 s
[1979:1979:0612/221632.507544:VERBOSE2:permission_rule_util.cc(18)] Browser window is not active

```	

</details>

**Pull request merged**